src/parallel/communicator.C

```diff
@@ -197,6 +197,19 @@ void Communicator::barrier () const
 void Communicator::barrier () const {}
 #endif
 
+#ifdef LIBMESH_HAVE_MPI
+void Communicator::nonblocking_barrier (Request & req) const
+{
+  if (this->size() > 1)
+    {
+      LOG_SCOPE("nonblocking_barrier()", "Parallel");
+      libmesh_call_mpi(MPI_Ibarrier (this->get(), req.get()));
```

Comment:
MPI_Ibarrier is an MPI3-only method, isn't it?  We almost certainly want to keep MPI-2 capability as a fallback, even if builds with MPI-3 work faster.  Didn't the code you showed me before use non-blocking reductions rather than non-blocking barriers?

Reply:
It did - but this is more efficient.  I do believe that it's MPI-3.

Why do we want to keep MPI-2?  MPI-2 is CRAZY old at this point.  Just like we moved on to C++11... we should move on.

Reply:
MPI-3 is 2012 BTW... and it's WAY easy to upgrade MPI